client/src/controllers/ActionController.test.js

```diff
@@ -86,6 +134,113 @@ describe('ActionController', () => {
     });
   });
 
+  describe('reload method', () => {
+    beforeEach(async () => {
+      await setup(`
+      <button
+        id="button"
+        data-controller="w-action"
+        data-action="click->w-action#reload"
+      >
+        Reload
+      </button>`);
+    });
+
+    it('should reload the page', () => {
+      const beforeUnloadHandler = jest.fn();
+      window.addEventListener('beforeunload', beforeUnloadHandler);
+
+      document.getElementById('button').click();
+
+      expect(window.location.reload).toHaveBeenCalledTimes(1);
+      expect(beforeUnloadHandler).toHaveBeenCalledTimes(1);
+
+      const event = beforeUnloadHandler.mock.lastCall[0];
+      // These mean the browser confirmation dialog was not shown
+      expect(event.defaultPrevented).toBe(false);
+      expect(event.returnValue).toBeNull();
+
+      window.removeEventListener('beforeunload', beforeUnloadHandler);
+    });
+
+    it('should not bypass the browser confirmation dialog if the event is prevented', async () => {
+      document.body.innerHTML = /* html */ `
+      <form
+        data-controller="w-unsaved"
+        data-action="beforeunload@window->w-unsaved#confirm"
+        data-w-unsaved-force-value="true"
+        data-w-unsaved-confirmation-value="You have unsaved changes!"
+      >
+      </form>
+      <button
+        id="button"
+        data-controller="w-action"
+        data-action="click->w-action#reload"
+      >
+        Reload
+      </button>
+      `;
+      await Promise.resolve();
+      const beforeUnloadHandler = jest.fn();
+      window.addEventListener('beforeunload', beforeUnloadHandler);
+
+      document.getElementById('button').click();
+
+      expect(window.location.reload).toHaveBeenCalledTimes(1);
+      expect(beforeUnloadHandler).toHaveBeenCalledTimes(1);
+
+      const event = beforeUnloadHandler.mock.lastCall[0];
+      // This means the browser confirmation dialog was shown
+      expect(event.returnValue).toBe('You have unsaved changes!');
+      window.removeEventListener('beforeunload', beforeUnloadHandler);
+    });
+  });
+
+  describe('forceReload method', () => {
+    beforeEach(async () => {
+      await setup(/* html */ `
+      <form
+        data-controller="w-unsaved"
+        data-action="beforeunload@window->w-unsaved#confirm"
+        data-w-unsaved-force-value="true"
+        data-w-unsaved-confirmation-value="You have unsaved changes!"
+      >
+      </form>
+      <button
+        id="button"
+        data-controller="w-action"
+        data-action="click->w-action#forceReload"
+      >
+        Force reload
+      </button>`);
+    });
+
+    it('should reload the page without preventing the beforeunload event', () => {
```

Comment:
`force-value` comes from `UnsavedController` and you are half-correct: it's to force the controller to think that there are unsaved changes. In real life, this is used after submitting an invalid form, so you immediately have the "unsaved changes" message at the bottom and the browser confirmation will be shown even if you don't make further edits since the page load (as it assumes your invalid changes are not yet saved).

Even when `force-value` is `true`, there's a legit use case where you want to show the "unsaved changes" message at the footer but also bypass the browser confirmation: when you submit the form. This is done by setting `confirmationValue` to an empty string:

https://github.com/wagtail/wagtail/blob/2d568dd825e4b087d7cea1c4d1fa49d916b395c2/client/src/controllers/UnsavedController.ts#L234-L241

The actual trigger is here:
https://github.com/wagtail/wagtail/blob/2d568dd825e4b087d7cea1c4d1fa49d916b395c2/client/src/controllers/UnsavedController.ts#L174-L191

You can see that `forceValue` basically is an override for `hasCommentsValue` and `hasEditsValue`, but it doesn't have complete control over whether the browser confirmation dialog is shown or not (as the `confirmationValue` takes priority). By the way, the way we trigger the browser confirmation really should be reworked (#12132).

TLDR: I'm just using `force-value` to avoid having to set up the "has edits" detection. I tried setting `has-edits-value` to `true` directly on the element, but apparently it clears it on `connect()` so that didn't work.

https://github.com/wagtail/wagtail/blob/2d568dd825e4b087d7cea1c4d1fa49d916b395c2/client/src/controllers/UnsavedController.ts#L98-L99

https://github.com/wagtail/wagtail/blob/2d568dd825e4b087d7cea1c4d1fa49d916b395c2/client/src/controllers/UnsavedController.ts#L166-L172

I guess I can try setting it after `connect()`, by doing `await Promise.resolve()` first.

Reply:
OK I've changed the test to use has-edits instead, and added a code comment as well.